### PR TITLE
Update validation.ts to differentiate between required and optional questions.

### LINKED ISSUE #1418: JS question validation only validates required questions, or attempted optional questions.
**Is your feature request related to a problem? Please describe.**
If a question is optional and I leave it blank, I should not see or be blocked by validation errors.

**Describe the solution you'd like**
JS question validation only applies to:
1. all required questions
2. optional questions with at least one field with a value

**Additional context**
#1416 introduces a class which can be used by js validation.

**Done when**
What conditions need to be met before this issue can be closed?

### PR DESCRIPTION
**Pull request by AWBaumann**
### Description
Empty optional questions are valid. Otherwise they need to meet the question's validation criteria.

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #1418

### CODE REVIEW

File: universal-application-tool-0.0.1/app/assets/javascripts/validation.ts [hunk 1]
```diff
@@ -196,31 +197,42 @@ class ValidationController {
...
 
       // validate zip code.
       const zipCode = <HTMLInputElement>question.querySelector(".cf-address-zip input");
+      const zipEmpty = zipCode.value.length == 0;
       const hasValidZip = zipCode.value.length == 5 && /^\d+$/.test(zipCode.value);
       // Change styling of '.cf-address-zip-error' as necessary.
       this.updateFieldErrorState(question, '.cf-address-zip', hasValidZip);
 
-      const hasEmptyInputs = !(addressLine1Valid && cityValid && stateValid && zipCode.value.length > 0);
+      const hasEmptyInputs = addressLine1Empty || cityEmpty || stateEmpty || zipEmpty;
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/assets/javascripts/validation.ts:229]: praise: thanks for simplifying this expression.

File: universal-application-tool-0.0.1/app/assets/javascripts/validation.ts [hunk 2]
```diff
@@ -296,15 +308,23 @@ class ValidationController {
     for (const question of nameQuestions) {
       // validate first name is not empty.
       const firstNameInput = <HTMLInputElement>question.querySelector(".cf-name-first input");
-      const firstNameValid = firstNameInput.value.length > 0;
-      this.updateFieldErrorState(question, '.cf-name-first', firstNameValid);
+      const firstNameEmpty = firstNameInput.value.length > 0;
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/assets/javascripts/validation.ts:311]:
```suggestion
      const firstNameEmpty = firstNameInput.value.length == 0;

```

File: universal-application-tool-0.0.1/app/assets/javascripts/validation.ts [hunk 3]
```diff
@@ -296,15 +308,23 @@ class ValidationController {
...
+      this.updateFieldErrorState(question, '.cf-name-first', !firstNameEmpty);
 
       // validate last name is not empty.
       const lastNameInput = <HTMLInputElement>question.querySelector(".cf-name-last input");
-      const lastNameValid = lastNameInput.value.length > 0;
-      this.updateFieldErrorState(question, '.cf-name-last', lastNameValid);
+      const lastNameEmpty = lastNameInput.value.length == 0;
+      this.updateFieldErrorState(question, '.cf-name-last', !lastNameEmpty);
+
+      // validate last name is not empty.
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/assets/javascripts/validation.ts:319]:
```suggestion
      // check if middle name is empty.

```

File: universal-application-tool-0.0.1/app/assets/javascripts/validation.ts [hunk 4]
```diff
@@ -296,15 +308,23 @@ class ValidationController {
...
+      const middleNameInput = <HTMLInputElement>question.querySelector(".cf-name-middle input");
+      const middleNameEmpty = middleNameInput.value.length == 0;
+
+      // If this question isn't required then it's also valid if it is empty.
+      const isOptional = !question.classList.contains(ValidationController.REQUIRED_QUESTION_CLASS);
+      const emptyOptional = isOptional && (!firstNameEmpty && !lastNameEmpty && middleNameEmpty);
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/assets/javascripts/validation.ts:325]:
```suggestion
      const emptyOptional = isOptional && firstNameEmpty && lastNameEmpty && middleNameEmpty;

```

File: universal-application-tool-0.0.1/app/assets/javascripts/validation.ts [hunk 5]
```diff
@@ -296,15 +308,23 @@ class ValidationController {
...
 
-      isValid = firstNameValid && lastNameValid;
+      isValid = emptyOptional || !(firstNameEmpty || lastNameEmpty);
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/assets/javascripts/validation.ts:327]:
```suggestion
      isValid = emptyOptional || (!firstNameEmpty && !lastNameEmpty);

```

Comment (hmhsu) [reply universal-application-tool-0.0.1/app/assets/javascripts/validation.ts:327]: Suggestion (non-blocking): this is equivalent, but I feel this is easier to reason about, for me at least.

**Review by hmhsu — COMMENTED** [review]
Looks good, with a few comments.

File: universal-application-tool-0.0.1/app/assets/javascripts/validation.ts (hunk 2 above)
Comment (AWBaumann) [reply universal-application-tool-0.0.1/app/assets/javascripts/validation.ts]: Done.

**Review by hmhsu — APPROVED** [review]

**Pull request merged**